raiden/network/proxies/token_network.py

```diff
@@ -812,96 +846,168 @@ def _set_total_deposit(
                 partner=partner,
             )
             self.client.poll(transaction_hash)
-            receipt_or_none = check_transaction_threw(self.client, transaction_hash)
+            receipt = check_transaction_threw(self.client, transaction_hash)
 
-        transaction_executed = gas_limit is not None
-        if not transaction_executed or receipt_or_none:
-            if transaction_executed:
-                block = receipt_or_none["blockNumber"]
-            else:
-                block = checking_block
+            if receipt:
+                # Because the gas estimation succeeded it is known that:
+                # - The channel was open.
+                # - The account had enough tokens to deposit
+                # - The account had enough balance to pay for the gas (however
+                #   there is a race condition for multiple transactions #3890)
+                failed_at_blockhash = encode_hex(receipt["blockHash"])
+
+                if receipt["cumulativeGasUsed"] == gas_limit:
+                    msg = (
+                        f"setTotalDeposit failed and all gas was used "
+                        f"({gas_limit}). Estimate gas may have underestimated "
+                        f"setTotalDeposit, or succeeded even though an assert is "
+                        f"triggered, or the smart contract code has an "
+                        f"conditional assert."
+                    )
+                    raise RaidenUnrecoverableError(msg)
+
+                # Query the current state to check for transaction races
+                our_details = self._detail_participant(
+                    channel_identifier=channel_identifier,
+                    detail_for=self.node_address,
+                    partner=partner,
+                    block_identifier=failed_at_blockhash,
+                )
+                partner_details = self._detail_participant(
+                    channel_identifier=channel_identifier,
+                    detail_for=self.node_address,
+                    partner=partner,
+                    block_identifier=failed_at_blockhash,
+                )
+
+                channel_state = self._get_channel_state(
+                    participant1=self.node_address,
+                    participant2=partner,
+                    block_identifier=failed_at_blockhash,
+                    channel_identifier=channel_identifier,
+                )
+                # Check if deposit is being made on a nonexistent channel
+                if channel_state != ChannelState.OPENED:
+                    msg = "Deposit failed as the channel was not open"
+                    raise RaidenUnrecoverableError(msg)
+
+                deposit_amount = total_deposit - our_details.deposit
+
+                total_channel_deposit = total_deposit + partner_details.deposit
+                if total_channel_deposit >= UINT256_MAX:
+                    raise RaidenUnrecoverableError("Deposit overflow")
```

Comment:
This should be recoverable. This node can race with the partner to do a deposit.

Additionally, could you please add a comment here, roughly:

> If an overflow is possible then we are interacting with a bad token. This must **not** crash the client, because it is not a Raiden bug, and otherwise this could be an attack vector.